<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [ ]:
!pip install psycopg2

In [ ]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = ''
PASSWORD = ''
HOST = ''
PORT = 5432


In [ ]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [ ]:
# код запроса представляет из себя строковую переменную
n = 10
query = f'''select *
           from vacancies
           limit {n}
        '''
# лучше использовать f-строки — так удобнее передавать параметры

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query, connection)
df

<ipython-input-32-3d34526774c1>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   
5  55906072                                          Корректор   
6  55523207                   React Front-end Middle Developer   
7  55050261                                Инженер-программист   
8  55583432                            Системный администратор   
9  55312414        Ведущий специалист группы разработки систем   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   
5  Корректура текстов\tГрамотность\tРабота в кома...       Полный день   
6       JavaScript\tTypeScript\tCSS3\tNode.js\tHTML5     Гибкий график   
7  Java SE\tSpring Framework\tSpring Boot\tGit\tS...       Полный день   
8  Linux\tАдминистрирование сетевого оборудования...       Полный день   
9  1С программирование\t1С: Предприятие 8\t1С: До...       Полный день   

           experience           employment  salary_from  salary_to  area_id  \
0           Нет опыта     Полная занятость      64000.0        NaN     1516   
1       От 3 до 6 лет     Полная занятость          NaN        NaN      130   
2       От 3 до 6 лет     Полная занятость          NaN        NaN      160   
3       От 3 до 6 лет     Полная занятость          NaN        NaN      159   
4           Нет опыта  Частичная занятость      15000.0        NaN     1955   
5           Нет опыта     Полная занятость          NaN        NaN     2323   
6  От 1 года до 3 лет     Полная занятость          NaN        NaN     2759   
7           Нет опыта     Полная занятость      40000.0    60000.0       49   
8  От 1 года до 3 лет     Полная занятость      50000.0        NaN        3   
9       От 3 до 6 лет     Полная занятость          NaN        NaN       47   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740  
5       584934  
6       204511  
7      4448636  
8        55126  
9      1413754

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [ ]:
# текст запроса
query_3_1 = f'''select count(id)
                from vacancies
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_3_1, connection)
df

<ipython-input-34-b7a408d7eafb>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

In [ ]:
# результат запроса
# В таблице данных по вакансиям 49197 записи

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [ ]:
# текст запроса
query_3_2 = f''' select count(id)
                 from employers
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_3_2, connection)
df

<ipython-input-36-4dc535f46d13>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

In [ ]:
# результат запроса
# В таблице данных по работодалеям 23501 записи

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [ ]:
# текст запроса
query_3_3 = f''' select count(id)
                 from areas
'''


In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_3_3, connection)
df

<ipython-input-42-2546ad89c83a>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

In [ ]:
# результат запроса
# В таблице кодов регионов и их названий 1362 записи

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [ ]:
# текст запроса
query_3_4 = f''' select count(id)
                 from industries
'''


In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_3_4, connection)
df

<ipython-input-55-26a36073868e>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

In [ ]:
# результат запроса
# В таблице вариантов сфер деятельности работодателей 294 записи

***

In [ ]:
# выводы по предварительному анализу данных


Исходя из предварительного анализа данных можем сказать следующее:
- количество вакансий в базе данных - 49197
- количество работодателей - 23501
- количество сфер деятельности - 294
Таким образом, можно сделать вывод о достаточно большом пуле данных.
Дальнейшие выводы будем делать исходя из детального анализа.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query_4_1 = f'''
SELECT
    a.name AS area,
    COUNT(v.id) AS cnt
FROM
    areas a
LEFT JOIN
    vacancies v ON a.id = v.area_id
GROUP BY
    a.name
ORDER BY
    cnt DESC;
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_4_1, connection)
df

<ipython-input-76-b1f80676e2db>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0                                Москва  5333
1                       Санкт-Петербург  2851
2                                 Минск  2112
3                           Новосибирск  2006
4                                Алматы  1892
...                                 ...   ...
1357                         Багаевская     0
1358                         Пролетарск     0
1359                       Седельниково     0
1360                             Инской     0
1361  Белоярский (Свердловская область)     0

[1362 rows x 2 columns]

In [ ]:
# результат запроса
# пятёрка лидеров по количеству вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''
    SELECT
    COUNT(*) AS cnt
FROM
    vacancies
WHERE
    salary_from IS NOT NULL
    OR salary_to IS NOT NULL;
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_4_2, connection)
df

<ipython-input-80-c00fae1e5f32>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

In [ ]:
# результат запроса
# У 24073 вакансий заполнено хотя бы одно из двух полей с зарплатой

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [ ]:
# текст запроса
query_4_3 = f'''
SELECT
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to
FROM
    vacancies
WHERE
    salary_from IS NOT NULL
    OR salary_to IS NOT NULL;
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_4_3, connection)
df

<ipython-input-86-3cf9dfcc5949>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

In [ ]:
# результат запроса
# получили средние значения нижней и верхней границ зарплатной вилки

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''
SELECT
    schedule,
    employment,
    COUNT(*) AS cnt
FROM
    vacancies
GROUP BY
    schedule, employment
ORDER BY
    cnt DESC;
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_4_4, connection)
df

<ipython-input-94-ffd63740b66c>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

In [ ]:
# результат запроса
# получили таблицу сочетания графика и трудоустройства, отсортированную в порядке убывания

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [ ]:
# текст запроса
query_4_5 = f'''
SELECT
  COUNT(*) AS cnt,
  experience
FROM
  vacancies
GROUP BY experience
ORDER BY cnt;
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_4_5, connection)
df

<ipython-input-110-050f5d84c456>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


cnt          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

In [ ]:
# результат запроса
# в результате запроса получили таблицу с количеством вакансий в зависимости от опыта работы. Наибольшее количество вакансий приходится на опыт работы от 1 до 3 лет.

***

In [ ]:
# выводы по детальному анализу вакансий
Исходя из детального анализа вакансий можем сказать следующее:
- распределение вакансий неравномерное: наиболеее число вакансий в наиболее крупных городах (Москва, Санкт-Петербург, Минск, Новосибирск, Алматы),
- заработную плату в вакансиях указывают только в приблизительно половине вакансий: всего 24073 вакансии (из 49197) содержат хотя бы одно значение заработной платы,
- средняя вилка заработной платы  - от 71 до 110 тыс. руб.,
- самое востребованное сочетание режима и графика работы - полный день и полная занятость,
- наибольшее число вакансий для сотрудников с опытом от 1 до 3 лет.

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query_5_1 = f'''
SELECT
    e.name AS employer,
    COUNT(v.id) AS cnt
FROM
    employers e
LEFT JOIN
    vacancies v ON e.id = v.employer_id
GROUP BY
    e.name
ORDER BY
    cnt DESC
LIMIT 10;
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_1, connection)
df

<ipython-input-118-024f7e0b27a4>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


employer   cnt
0                 Яндекс  1933
1             Ростелеком   491
2               Тинькофф   444
3                   СБЕР   428
4          Газпром нефть   331
5            ИК СИБИНТЕК   327
6                    МТС   292
7                DataArt   247
8  Совкомбанк Технологии   204
9             Первый Бит   176

In [ ]:
# результат запроса
# в результате запроса получилась таблица, отсротированная по убыванию, с количеством вакансии в разрезе работодателей

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
# Запрос на получение количества работодателей и вакансий по каждому региону
query_5_2_1 = f'''
SELECT
    a.id AS area_id,
    a.name AS area_name,
    COUNT(distinct e.id) AS employer_count,
    COUNT(v.id) AS vacancy_count
FROM
    areas a
LEFT JOIN
    employers e ON a.id = e.area
LEFT JOIN
    vacancies v ON a.id = v.area_id
GROUP BY
    a.id, a.name
ORDER BY
    employer_count DESC;
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_2_1, connection)
df

<ipython-input-7-60056f5a41b8>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2_1, connection)


area_id                            area_name  employer_count  \
0           1                               Москва            5864   
1           2                      Санкт-Петербург            2217   
2        1002                                Минск            1115   
3         160                               Алматы             721   
4           3                         Екатеринбург             609   
...       ...                                  ...             ...   
1357     5461                              Янино-1               0   
1358     5485                          Адыге-Хабль               0   
1359     5504                          Эркен-Шахар               0   
1360     5524  Индустриальный (Краснодарский край)               0   
1361     5545          Сабетта (Ямало-Ненецкий АО)               0   

      vacancy_count  
0          31272712  
1           6320667  
2           2354880  
3           1364132  
4           1034082  
...             ...  
1357              0  
1358              1  
1359              0  
1360              0  
1361              0  

[1362 rows x 4 columns]

In [ ]:
# текст запроса
# Запрос на получение количества работодателей и вакансий по в случае отсутствия региона
query_5_2_2 = f'''
SELECT areas.name,
       count(employers.id) AS COUNT
FROM areas
LEFT JOIN employers ON areas.id = employers.area
WHERE areas.id NOT IN
    (SELECT area_id
     FROM vacancies)
GROUP BY areas.name
ORDER BY COUNT DESC
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_2_2, connection)
df

<ipython-input-8-3a6c6f93da63>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2_2, connection)


name  count
0                Россия    410
1             Казахстан    207
2    Московская область     75
3    Краснодарский край     19
4    Ростовская область     18
..                  ...    ...
588            Новоорск      0
589              Алупка      0
590     Новозавидовский      0
591               Бирюч      0
592           Енотаевка      0

[593 rows x 2 columns]

In [ ]:
# результат запроса
# Наибольшее количество вакансий в Москве, наибольшее количесво вакансий без указания региона - в России.

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''
SELECT
    e.name AS employer_name,
    COUNT(DISTINCT v.area_id) AS regions_cnt
FROM
    employers e
LEFT JOIN
    vacancies v ON e.id = v.employer_id
GROUP BY
    e.id, e.name
ORDER BY
    regions_cnt DESC, e.name;
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_3, connection)
df

<ipython-input-10-c204f9ddf8e5>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer_name  regions_cnt
0                       Яндекс          181
1                   Ростелеком          152
2                   Спецремонт          116
3       Поляков Денис Иванович           88
4                    ООО ЕФИН            71
...                        ...          ...
23496             Ярпож Казань            0
23497                   ЯРСНИП            0
23498   Ясли-сад № 28 г.Минска            0
23499            Яшин&Партнёры            0
23500  Яшнова Алена Викторовна            0

[23501 rows x 2 columns]

In [ ]:
# результат запроса
# посчитали количество регионов, в котрых опубликованы вакансии работодателей. Лидер списка - Яндекс

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [ ]:
# текст запроса
query_5_4 = f'''
SELECT
    COUNT(*) AS employers_without_industry
FROM
    employers e
LEFT JOIN
    employers_industries ei ON e.id = ei.employer_id
WHERE
    ei.employer_id IS NULL;
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_4, connection)
df

<ipython-input-12-fc7fb6063681>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employers_without_industry
0                        8419

In [ ]:
# результат запроса
# в результате запроса мы получили количество работодалетей, которые не указали сферу деятельности в вакансии

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [ ]:
# текст запроса
query_5_5 = f'''
WITH employers_with_four_industries AS (
    SELECT
        e.name AS employer_name
    FROM
        employers e
    JOIN
        employers_industries ei ON e.id = ei.employer_id
    GROUP BY
        e.id, e.name
    HAVING
        COUNT(ei.industry_id) = 4
)
SELECT
    employer_name
FROM
    employers_with_four_industries
ORDER BY
    employer_name ASC
LIMIT 1 OFFSET 2;
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_5, connection)
df

<ipython-input-14-09c078ca4ce8>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


employer_name
0          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [ ]:
# текст запроса
query_5_6 = f'''
SELECT
    COUNT(DISTINCT e.id) AS employers_count
FROM
    employers e
JOIN
    employers_industries ei ON e.id = ei.employer_id
JOIN
    industries i ON ei.industry_id = i.id
WHERE
    i.name = 'Разработка программного обеспечения';
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_6, connection)
df

<ipython-input-16-f6bffdcc7bb1>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


employers_count
0             3553

In [ ]:
# результат запроса
# В базе 3553 работодателя со сферой деятельности "Разработка программного обеспечения"

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [ ]:
# код для получения списка городов-милионников
query_5_7 = f'''
   SELECT employers.name,
       areas.name,
       count(areas.name)
FROM vacancies
LEFT JOIN employers ON employer_id = employers.id
LEFT JOIN areas ON area_id = areas.id
GROUP BY employers.name,
         areas.name
HAVING employers.name = 'Яндекс'
AND areas.name IN ('Москва',
                   'Санкт-Петербург',
                   'Новосибирск',
                   'Екатеринбург',
                   'Казань',
                   'Красноярск',
                   'Нижний Новгород',
                   'Челябинск',
                   'Уфа',
                   'Самара',
                   'Ростов-на-Дону',
                   'Краснодар',
                   'Омск',
                   'Воронеж',
                   'Пермь',
                   'Волгоград')
UNION
SELECT 'Total',
       '',
       count(employers.name)
FROM vacancies
LEFT JOIN employers ON employer_id = employers.id
LEFT JOIN areas ON area_id = areas.id
WHERE employers.name = 'Яндекс'
  AND areas.name IN ('Москва',
                     'Санкт-Петербург',
                     'Новосибирск',
                     'Екатеринбург',
                     'Казань',
                     'Красноярск',
                     'Нижний Новгород',
                     'Челябинск',
                     'Уфа',
                     'Самара',
                     'Ростов-на-Дону',
                     'Краснодар',
                     'Омск',
                     'Воронеж',
                     'Пермь',
                     'Волгоград')
ORDER BY COUNT ASC
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_5_7, connection)
df

<ipython-input-22-fc06feddb56d>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name             name  count
0   Яндекс             Омск     21
1   Яндекс        Челябинск     22
2   Яндекс       Красноярск     23
3   Яндекс        Волгоград     24
4   Яндекс            Пермь     25
5   Яндекс           Казань     25
6   Яндекс   Ростов-на-Дону     25
7   Яндекс              Уфа     26
8   Яндекс           Самара     26
9   Яндекс        Краснодар     30
10  Яндекс          Воронеж     32
11  Яндекс      Новосибирск     35
12  Яндекс  Нижний Новгород     36
13  Яндекс     Екатеринбург     39
14  Яндекс  Санкт-Петербург     42
15  Яндекс           Москва     54
16   Total                     485

In [ ]:
# результат запроса
# В результате запроса получили список количества вакансий Яндекса по городам милионникам

***

In [ ]:
# выводы по анализу работодателей


Исходя из детального анализа работодателей можем сделать следующие выводы:
- наибольшее количество вакансии - у крупных игроков рынка (Яндекс, Ростелеком, Тинькофф, Сбер и др)
- регионы-лидеру по числу вакансий - Москва, Санкт-Петербург, Микс, Алматы, Новосибирск
- наибольшее количество регионов, в которых открыты вакансии у таких крупных компаний, как Яндекс, Ростелеком, Спецремонт и некий Поляков Денис Иванович :)
- в базе данных 8419 работодателей, не указавших свою сферу деятельности,
- в базе 3553 компании-разработчика ПО.

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [ ]:
# текст запроса
query_6_1 = f'''
SELECT
    COUNT(*) AS data_related_vacancies
FROM
    vacancies
WHERE
    LOWER(name) LIKE '%data%'
    OR LOWER(name) LIKE '%данн%';
    '''

In [ ]:

# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_6_1, connection)
df

<ipython-input-17-8ff419cbf18a>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


data_related_vacancies
0                    1771

In [ ]:
# результат запроса

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [ ]:
# текст запроса
query_6_2 = f'''
SELECT
    COUNT(*) AS junior_data_scientist_vacancies
FROM
    vacancies
WHERE
    (LOWER(name) LIKE '%data scientist%'
    OR LOWER(name) LIKE '%data science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%')
    AND (
        LOWER(name) LIKE '%junior%'
        OR LOWER(experience) = 'нет опыта'
        OR LOWER(employment) = 'стажировка'
    );
    '''

In [ ]:

# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_6_2, connection)
df

<ipython-input-20-a4d02860e3d4>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


junior_data_scientist_vacancies
0                               51

In [ ]:
# результат запроса

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = f'''
SELECT
    COUNT(*) AS ds_vacancies_with_sql_or_postgres
FROM
    vacancies
WHERE
    (LOWER(name) LIKE '%data scientist%'
    OR LOWER(name) LIKE '%data science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%')
    AND (LOWER(key_skills) LIKE '%sql%'
    OR LOWER(key_skills) LIKE '%postgres%');
    '''

In [ ]:

# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_6_3, connection)
df

<ipython-input-22-656f7baed261>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


ds_vacancies_with_sql_or_postgres
0                                229

In [ ]:
# результат запроса

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = f'''
SELECT
    COUNT(*) AS ds_vacancies_with_python
FROM
    vacancies
WHERE
    (LOWER(name) LIKE '%data scientist%'
    OR LOWER(name) LIKE '%data science%'
    OR LOWER(name) LIKE '%исследователь данных%'
    OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
    OR LOWER(name) LIKE '%machine learning%'
    OR LOWER(name) LIKE '%машинн%обучен%')
    AND LOWER(key_skills) LIKE '%python%';
    '''

In [ ]:

# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_6_4, connection)
df

<ipython-input-24-509a0813de1b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


ds_vacancies_with_python
0                       357

In [ ]:
# результат запроса

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [ ]:
# текст запроса
query_6_5 = f'''
SELECT round(avg(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2)
FROM vacancies
WHERE (LOWER(name) NOT LIKE '%html%')
  AND (LOWER(name) NOT LIKE '%teamlead%')
  AND (LOWER(name) NOT LIKE '%qml%')/*and (LOWER(name) NOT LIKE '%aml%') and (LOWER(name) NOT LIKE '%joomla%')*/
  AND (LOWER(name) LIKE '%data scientist%'
       OR LOWER(name) LIKE '%data science%'
       OR LOWER(name) LIKE '%исследователь данных%'
       OR LOWER(name) LIKE '%ml%'
       OR LOWER(name) LIKE '%machine learning%'
       OR LOWER(name) LIKE '%машинн%обучен%')
  AND key_skills IS NOT NULL;
    '''

In [ ]:

# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_6_5, connection)
df

<ipython-input-25-750695bf16f4>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

In [ ]:
# результат запроса
# В среднем требуется не менее 6 навыков для специалистов Data Science

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [ ]:
# текст запроса
query_6_6 = f'''
SELECT experience,
       round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from)) / 2.0), 0) AS avg_salary
FROM vacancies
WHERE (salary_from IS NOT NULL
       OR salary_to IS NOT NULL)
  AND ((LOWER(name) NOT LIKE '%html%')
       AND (LOWER(name) NOT LIKE '%teamlead%')
       AND (LOWER(name) LIKE '%data scientist%'
            OR LOWER(name) LIKE '%data science%'
            OR LOWER(name) LIKE '%исследователь данных%'
            OR LOWER(name) LIKE '%ml%'
            OR LOWER(name) LIKE '%machine learning%'
            OR LOWER(name) LIKE '%машинн%обучен%'))
GROUP BY experience;
    '''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_6_6, connection)
df

<ipython-input-27-90fb779a0208>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    138307.0
2       От 3 до 6 лет    243115.0

In [ ]:
connection.close()

In [ ]:
# результат запроса
# Специалист DS с опытом работы от 3 до 6 лет в среднем получает 243 115 руб. в месяц

***

In [ ]:
# выводы по предметному анализу

Исходя из предметного анализа можем сделать следующие выводы:
- в базе 1771 вакансия, имеющая отношение к данным,
- при этом для начинающих специалистов 51 по данным вакансия,
- владение SQL - востребованный навык (присутствует в требованиях 229 вакансий),
- владение Python еще более вотребованный навык (357 вакансий предполагают владение этим ЯП),
- специалист по даннм должен иметь по крайне мере 6 ключевых навыков для успешного трудоустройства,
- наибольшая зарплата у специалистов с большим опытом.

## Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

В завершении нашего исследования можем сделать следующие выводы.
Рынок труда испытывает дефицит квалифицированных кадров. Наибольшее количество вакансий - у крупных компаний и в крупных городах.
Специалисты по данным востребованы на рынке труда. При этом наибольшим спросом пользуются специалисты с опытом и обширным пулом навыков.